### ISSUE TITLE
banlist processing in core or plugins?

### ISSUE THREAD

**Issue by ffries** [opnsense/core#1836]
Dear all,

I would iike to point out this banlist source:
https://raw.githubusercontent.com/openwrt/packages/lede-17.01/net/adblock/files/adblock.conf

From my memory, it contains around 140.000 entries.

It appears that most banlists need to be processed using regex/php before being saved to an IP list.
Example: 'securemecca'
http://securemecca.com/Downloads/hosts.txt'
'\$0 ~/^127\.0\.0\.1[ \t]+([A-Za-z0-9_-]+\.){1,}[A-Za-z]+/{print tolower(\$2)}'

There are 3 places where banlists could be used:
* pf table firewall
* proxy banlist
* dns banlist

And maybe more (intrusion detection?)

So IMHO we need some kind of banlist processing feature inside OPNsense.
In OPNsense, I see 2 implementation solutions:

* implement banlist processing in core.
Basically for pf table and proxy.
Add a regex entry and process the regex after downloading.

* implement banlist processing in a separate plugin.
port adblock2 to an OPNsense plugin: download, filter, merge, whitelist, blacklist, one or several files ...

The resulting files could then be used in pf table, proxy and dns banlists and would be more flexible.

What are your plans as regards banlist processing?

**Comment by fichtner** [comment]
Hi,

This easily reaches meta-management level of lists: their origin, purpose, their format, their conversion, schedules of updates and internal distribution to several services with different expectations in output format.

The highest value would be import of lists into aliases to be used as addresses / networks or hostnames. At least in aliases, the update mechanism is well established and integrated, but then these can only be used for the firewall rules.

What's second-level scope is using aliases in the traffic shaper, possibly extending this to DNS servers. Third tier is proxy integration, whether by separate plugin or shared core functionality is a question to be answered after having gathered experience with the first and second level of such an endeavour. Traditionally, such functionality is not shared due to the nature of proxy blocklists, intrusion detection rules and simple address lists.

It's not much of a "plan" as it is a vision of how components interact and may share common mechanisms. The idea is to think big but start in tiny steps towards a goal and reassess after having learned something unexpected in the process. I hope this helps nevertheless.


Cheers,
Franco

**Comment by ffries** [comment]
You may have a look at https://raw.githubusercontent.com/openwrt/packages/lede-17.01/net/adblock
I think I am going to simply port it to a plugin.

**Comment by fichtner** [comment]
this is support for lack of a better label. "support" as in helping to get something to progress :)

the link is a 404 ?

**Comment by ffries** [comment]
Hello. 

The link used to work.
I am trying to find a suitable project to start OPNsense development :)

The original thread for adblock is:
https://forum.lede-project.org/t/adblock-support-thread/507

I hope that you don't mind that I put this URL here.
LEDE is Gnu/linux, so just plain of black holes and security holes.
This is why i migrated to FreeBSD/OpenBSD/OPNsense.

I will manage the adblock pluguin port to OPNsense.

**Comment by ffries** [comment]
Closing. Will submit a plugin when ready.
